tezos/encoding-derive/src/make.rs

```diff
@@ -0,0 +1,689 @@
+    Ok(encoding)
+}
+
+/// Consumes the rightmost meta attribute to make basic encoding.
+fn get_basic_encoding_from_meta<'a>(
+    _ty: &'a syn::Path,
+    meta: &mut Vec<syn::Meta>,
+) -> Result<Option<Encoding<'a>>> {
+    let encoding = if let Some(bytes) = get_attribute_no_param(meta, &symbol::STRING)? {
+        Encoding::Bytes(bytes.span)
+    } else if let Some(string) =
+        get_attribute_with_option(meta, &symbol::STRING, Some(&symbol::MAX), true)?
+    {
+        Encoding::String(string.param, string.span)
+    } else if let Some(zarith) = get_attribute_no_param(meta, &symbol::Z_ARITH)? {
+        Encoding::Z(zarith.span)
+    } else if let Some(mutez) = get_attribute_no_param(meta, &symbol::MU_TEZ)? {
+        Encoding::MuTez(mutez.span)
+    } else if let Some(builtin) =
+        get_attribute_with_param(meta, &symbol::BUILTIN, Some(&symbol::KIND), true)?
+    {
+        Encoding::Primitive(builtin.param, builtin.span)
+    } else {
+        return Ok(None);
+    };
+    Ok(Some(encoding))
+}
+
+/// Creates basic encoding by consumes the rightmost meta attribute.
+fn make_basic_encoding_from_meta<'a>(
+    ty: &'a syn::Path,
+    meta: &mut Vec<syn::Meta>,
+) -> Result<Encoding<'a>> {
+    get_basic_encoding_from_meta(ty, meta)?
+        .ok_or_else(|| error_spanned(ty, "No basic encoding specified"))
+}
+
+/// Returns content of the `composite` attribute.
+fn get_composite_meta(meta: &mut Vec<syn::Meta>) -> Result<Option<Vec<syn::Meta>>> {
+    if let Some(composite) = get_attribute(meta, &symbol::COMPOSITE) {
+        if let syn::Meta::List(list) = composite {
+            let list = list
+                .nested
+                .into_iter()
+                .map(|nested| {
+                    if let syn::NestedMeta::Meta(meta) = nested {
+                        Ok(meta)
+                    } else {
+                        Err(error_spanned(nested, "Invalid nested attribute type"))
+                    }
+                })
+                .collect::<Result<_>>()?;
+            Ok(Some(list))
+        } else {
+            Err(error_spanned(composite, "Invalid attribute type"))
+        }
+    } else {
+        Ok(None)
+    }
+}
+
+/// Asserts that meta attribute corresponds to the specified built-in encoding `kind`.
+fn assert_builtin_encoding(meta: &mut Vec<syn::Meta>, kind: &PrimitiveEncoding) -> Result<()> {
+    if let Some(builtin) = get_attribute_with_param::<PrimitiveEncoding>(
+        meta,
+        &symbol::BUILTIN,
+        Some(&symbol::KIND),
+        true,
+    )? {
+        if *kind != builtin.param {
+            return Err(error(
+                builtin.span,
+                "Built-in encoding does not match the type",
+            ));
+        }
+    } else if let Some(string) = get_attribute(meta, &symbol::STRING) {
+        return Err(error_spanned(
+            string,
+            "String encoding can be used only with `String` type",
+        ));
+    }
+    Ok(())
+}
+
+/*
```

Comment:
@akoptelov 
do we need this?
if not remove it please

Reply:
Removed.

Reply:
Removed.